src/pymor/operators/constructions.py

```diff
@@ -1430,3 +1430,97 @@ def apply_adjoint(self, V, mu=None):
     def apply_inverse_adjoint(self, U, mu=None, initial_guess=None, least_squares=False):
         assert U in self.source
         return self.range.from_numpy(U.to_numpy())
+
+
+class BilinearFunctional(Operator):
+    """An `Operator` intended as a bilinear functional.
+
+    Given an operator `A` acting on R^d, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> A(u, u)
+
+    In particular this is intended to provide the jacobian of such a bilinear
+    functional::
+
+        op.jacobian(U, mu).apply(V) = (A(u, v; mu) + A(v, u; mu))
+
+    Parameters
+    ----------
+    operator
+        The `Operator` to be wrapped as a bilinear functional.
+    """
+
+    linear = False
+    bilinear = True
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operator, name=None):
+        self.__auto_init(locals())
+        self.source = operator.source
+
+    @property
+    def H(self):
+        return BilinearFunctional(self.operator.H, name=self.name + '_adjoint')
+
+    def apply(self, U, mu=None):
+        assert U in self.source
+        return self.range.from_numpy(self.operator.apply2(U, U, mu))
+
+    def jacobian(self, U, mu=None):
+        inner_vec = self.operator.apply_adjoint(U, mu) + self.operator.apply(U, mu)
+        return VectorFunctional(inner_vec, name=self.name + '_jacobian')
+
+    def d_mu(self, parameter, index=1):
+        # the parameter derivative only takes effect on the inner operator
+        return BilinearFunctional(self.operator.d_mu(parameter, index), name=self.name + '_d_mu')
+
+
+class BilinearProductFunctional(Operator):
+    """An `Operator` intended as a bilinear functional by multiplying two linear functionals.
+
+    Given linear operators `A`, `B` acting on R^d, this class enables the functional::
+
+        op: R^d ----> R^1
+             u  |---> (A(u), B(u))
```

Comment:
From that I assume that A,B map to R^d, right?

Reply:
I was actually thinking about $K: \mathbb{R}^d \rightarrow \mathbb{R}^n$, I clarified that in the documentation now. This way it also makes sense to consider $(\cdot, \cdot)$, otherwise a simple scalar multiplicative would suffice.